### PR TITLE
[8381] add option to allow unregistered users to vote in polls

### PR DESCRIPTION
**Pull request by goapunk**
**Tasks**
- [x] PR name contains story or task reference
- [ ] Documentation (docs and inline)
- [x] Tests (including n+1 and django_assert_num_queries where applicable)
- [ ] Changelog

### CODE REVIEW

**Comment by goapunk** [comment]
@m4ra @hom3mad3 here's a version where everything is integrated into the existing poll module for comparison. It's less big of a change than I expected. No changes are necessary on the a+ side so can be tested with main.

File: adhocracy4/polls/api.py [hunk 1]
```diff
@@ -93,57 +100,108 @@ def retrieve(self, request, *args, **kwargs):
...
+                        user=self.request.user,
+                        organisation=self.get_object().project.organisation,
+                        defaults={
+                            "has_agreed": self.request.data["agreed_terms_of_use"]
+                        },
+                    )
             else:
                 raise ValidationError(
                     {_("Please agree to the organisation's terms of use.")}
                 )
```
Comment (m4ra) [inline adhocracy4/polls/api.py:154]: how the non-registered user will agree to the terms? This happens by default when they register IIRC

File: adhocracy4/polls/models.py [hunk 2]
```diff
@@ -164,7 +171,7 @@ def __str__(self):
 
     class Meta:
         ordering = ["id"]
-        unique_together = ("question", "creator")
+        unique_together = ("question", "creator", "content_id")
```
Comment (m4ra) [inline adhocracy4/polls/models.py:174]: does this work if either creator or content_id is None?

**Review by m4ra — COMMENTED** [review]
I left some comments for the django part, but need to do a local test too.
@hom3mad3 can you look more closely to the react part? 
BTW can we have the makefile and the github workflows in a separate PR?

File: adhocracy4/polls/models.py (hunk 2 above)
Comment (goapunk) [reply adhocracy4/polls/models.py:174]: no, None is treated as a normal value in the unique check as far as I understand, therefore e.g. `question1 + user1 + None` is a valid unique value. Same for `question1 + None + unique uuid` is a valid unique value. And both trigger an error if voted again with the same question and user/content_id

**Comment by goapunk** [comment]
> I left some comments for the django part, but need to do a local test too. @hom3mad3 can you look more closely to the react part? BTW can we have the makefile and the github workflows in a separate PR?

I'll make extra PRs but we will have to merge the workflow before this then and rebase it

File: adhocracy4/polls/serializers.py [hunk 3]
```diff
@@ -130,24 +137,24 @@ class PollSerializer(serializers.ModelSerializer):
...
                     ).count()
-                    + models.Answer.objects.filter(
-                        question__poll=poll, creator=user
-                    ).count()
+                    + Answer.objects.filter(question__poll=poll, creator=user).count()
                 ) > 0
         return False
 
     def update(self, instance, data):
+        instance.allow_unregistered_users = data.get("allow_unregistered_users", False)
```
[the remark this answers is not in the document]
Comment (goapunk) [reply adhocracy4/polls/serializers.py:156]: You mean to check whether the unregistered user trying to vote is allowed to do that ? That's done by the rules/predicates already.

**Comment by m4ra** [comment]
@goapunk tried to test it locally, but we cannot set the `allowed_unregistered_users` from the poll's dashboard yet. Maybe could be a checkbox in the basic settings? For registered users works as expected.

**Comment by goapunk** [comment]
> @goapunk tried to test it locally, but we cannot set the `allowed_unregistered_users` from the poll's dashboard yet. Maybe could be a checkbox in the basic settings? For registered users works as expected.

@m4ra it works for me, the checkbox is there (in the `Poll` setting). Did you do `npm link <path to adhocracy4>` ?

**Review by m4ra — APPROVED** [review]
LGTM, thank you!

**Pull request merged**